test/wasm/assets/test.js

```diff
@@ -54,19 +54,53 @@ function formatMicros(us) {
     }
 }
 
-function evaluate(mem, policy, input) {
+function loadJSON(mod, memory, value) {
 
-    const str = JSON.stringify(input)
-    const addr = policy.instance.exports.opa_malloc(str.length);
-    const buf = new Uint8Array(mem.buffer);
+    const str = JSON.stringify(value);
+    const rawAddr = mod.instance.exports.opa_malloc(str.length);
+    const buf = new Uint8Array(memory.buffer);
 
     for (let i = 0; i < str.length; i++) {
-        buf[addr + i] = str.charCodeAt(i);
+        buf[rawAddr + i] = str.charCodeAt(i);
     }
 
+    const parsedAddr = mod.instance.exports.opa_json_parse(rawAddr, str.length);
 
-    const returnCode = policy.instance.exports.eval(addr, str.length);
+    if (parsedAddr == 0) {
+        throw "failed to parse json value"
+    }
+
+    return parsedAddr;
+}
+
+async function instantiate(bytes, memory, data) {
+
+    const addr2string = stringDecoder(memory);
+
+    const mod = await WebAssembly.instantiate(bytes, {
+        env: {
+            memory: memory,
+            opa_abort: function (addr) {
+                throw addr2string(addr);
+            },
+            opa_println: function (addr) {
+                console.log(addr2string(addr));
+            },
+        },
+    });
 
+    const dataAddr = loadJSON(mod, memory, data);
+    const heapPtr = mod.instance.exports.opa_heap_ptr_get();
+    const heapTop = mod.instance.exports.opa_heap_top_get();
+
+    return { module: mod, memory: memory, heapPtr: heapPtr, heapTop: heapTop, dataAddr: dataAddr };
+}
+
+function evaluate(policy, input) {
+    policy.module.instance.exports.opa_heap_ptr_set(policy.heapPtr);
+    policy.module.instance.exports.opa_heap_top_set(policy.heapTop);
```

Comment:
Why do we get the heap ptr/top and then set them again right after? At a glance I'm not seeing why it needs to be done for each wasm module, mostly asking to clarify my understanding of this.

Reply:
Each time eval() is called we reset the addresses so that memory allocation proceeds from immediately after the section of memory where the cached data is loaded.

Reply:
Ohhh I missed the loop for each of the test cases and the shared memory. Makes sense 👍